crates/core/src/data_cache.rs

```diff
@@ -87,31 +94,48 @@ fn read_nil_padded<T>(
     filename: impl AsRef<camino::Utf8Path>,
 ) -> color_eyre::Result<Vec<T>>
 where
-    T: serde::de::DeserializeOwned,
+    T: for<'de> alox_48::Deserialize<'de>,
 {
     let path = camino::Utf8PathBuf::from("Data").join(filename);
     let data = filesystem.read(path)?;
 
+    let mut trace = alox_48::path_to_error::Trace::new();
     let mut de = alox_48::Deserializer::new(&data)?;
+    let de = alox_48::path_to_error::Deserializer::new(&mut de, &mut trace);
 
-    luminol_data::helpers::nil_padded::deserialize(&mut de).map_err(color_eyre::Report::from)
+    luminol_data::helpers::nil_padded_alox::deserialize_with(de)
+        .map_err(|error| format_traced_error(error, trace))
 }
 
 fn write_nil_padded(
-    data: &[impl serde::Serialize],
+    data: &[impl alox_48::Serialize],
     filesystem: &impl luminol_filesystem::FileSystem,
     filename: impl AsRef<camino::Utf8Path>,
 ) -> color_eyre::Result<()> {
     let path = camino::Utf8PathBuf::from("Data").join(filename);
 
+    let mut trace = alox_48::path_to_error::Trace::new();
     let mut ser = alox_48::Serializer::new();
+    let trace_ser = alox_48::path_to_error::Serializer::new(&mut ser, &mut trace);
 
-    luminol_data::helpers::nil_padded::serialize(data, &mut ser)?;
+    luminol_data::helpers::nil_padded_alox::serialize_with(data, trace_ser)
+        .map_err(|error| format_traced_error(error, trace))?;
     filesystem
         .write(path, ser.output)
         .map_err(color_eyre::Report::from)
 }
 
+fn format_traced_error(
+    error: impl std::fmt::Display,
+    trace: alox_48::path_to_error::Trace,
+) -> color_eyre::Report {
+    let mut message = format!("Error {error}:");
+    for context in trace.context {
+        write!(message, "\n  {}", context).unwrap();
+    }
+    color_eyre::Report::msg(message)
+}
```

Comment:
Why not use `color_eyre::eyre::eyre!` to create a report from the error message and then call `.wrap_err` on that report repeatedly to add the contexts from the trace to it? That would have the additional benefit of preserving the ability to determine the type of error by doing stuff like `report.root_cause().downcast_ref()` on the report later.